promql/engine.go

```diff
@@ -1531,18 +1531,22 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 			newEv.interval = ev.noStepSubqueryIntervalFn(rangeMillis)
 		}
 
-		// Start with the first timestamp after (ev.startTimestamp - offset - range)
-		// that is aligned with the step (multiple of 'newEv.interval').
-		newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
-		if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
-			newEv.startTimestamp += newEv.interval
-		}
-
-		if newEv.startTimestamp != ev.startTimestamp {
-			// Adjust the offset of selectors based on the new
-			// start time of the evaluator since the calculation
-			// of the offset with @ happens w.r.t. the start time.
-			setOffsetForAtModifier(newEv.startTimestamp, e.Expr)
+		if !e.AlignEvalTime {
+			// Start with the first timestamp after (ev.startTimestamp - offset - range)
+			// that is aligned with the step (multiple of 'newEv.interval').
+			newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
+			if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
+				newEv.startTimestamp += newEv.interval
+			}
+
+			if newEv.startTimestamp != ev.startTimestamp {
+				// Adjust the offset of selectors based on the new
+				// start time of the evaluator since the calculation
+				// of the offset with @ happens w.r.t. the start time.
+				setOffsetForAtModifier(newEv.startTimestamp, e.Expr)
+			}
+		} else {
+			newEv.startTimestamp = ev.startTimestamp - offsetMillis - rangeMillis + newEv.interval
```

Comment:
I guess `+ newEv.interval` is not required since it is still in the time range without that.

Additionally, we will need to call `setOffsetForAtModifier` for the same reasons as above.

Reply:
From my tests, without `+ newEv.interval` one less point is returned, eg. `[1m::15s]` with three points instead of four.

Reply:
Strange, that should be the opposite of what should be happening because `+` means we are moving start ahead, hence reducing the range. Anyway, we will know better once we have all the tests.

Reply:
With:

![image](https://user-images.githubusercontent.com/54278938/127419904-4d042428-976b-488f-b88a-2c5761da5180.png)

Without:

![image](https://user-images.githubusercontent.com/54278938/127419697-d71cd176-4727-4540-a8fa-5d164e5477c3.png)

Reply:
(side note: we'll have to update the linters)

Reply:
I checked at the screenshots, and the screenshot is actually correct without the +interval. Because, with `[1m::15s]`, you should get 5 samples because both the ends are inclusive. And since current subqueries are aligned with the step, almost always you will get 4 samples since the time ranges of the query will be not aligned with the step.

If you are getting 3 samples in the tests for `[1m::15s]`, can you verify if the raw data is correct?

Reply:
Ah, that makes sense. Thanks!

I can reproduce getting 3 samples with `prometheus_http_requests_total{...}[1m::15s]`, as well as 1, 2, and 4, but I think it's specific to that metric, which doesn't always have complete data. I can't reproduce it with any other, so I don't think it's an issue.